File: nafed_erp/procure_to_pay/api/stock_entry.py
import frappe

@frappe.whitelist()
def create_po_from_indent(stock_entry):

    ste = frappe.get_doc("Stock Entry", stock_entry)

    supplier = frappe.db.get_value("Supplier", {}, "name")

    if not supplier:
        frappe.throw("No Supplier found")

    po = frappe.new_doc("Purchase Order")
    po.supplier = supplier

    has_shortage = False

    for item in ste.items:

        requested_qty = getattr(item, "custom_requested_qty", None) or item.qty

        available_qty = frappe.db.get_value(
            "Bin",
            {
                "item_code": item.item_code,
                "warehouse": item.s_warehouse
            },
            "actual_qty"
        ) or 0

        # ✅ SAFE SHORTAGE
        shortage = max(0, requested_qty - available_qty)

        if shortage <= 0:
            continue

        has_shortage = True

        po.append("items", {
            "item_code": item.item_code,
            "qty": shortage,
            "schedule_date": frappe.utils.nowdate(),
            "warehouse": item.t_warehouse
        })

    if not has_shortage:
        frappe.throw("No shortage items to create PO")

    po.insert()
    po.submit()

    return po.name


def calculate_indent_data(doc, method=None):

    requested_qty = 0
    available_qty_total = 0

    fully_available = 0
    partially_available = 0
    not_available = 0

    requested_sku = len(doc.items)

    for item in doc.items:

        available_qty = frappe.db.get_value(
            "Bin",
            {
                "item_code": item.item_code,
                "warehouse": item.s_warehouse
            },
            "actual_qty"
        ) or 0

        req_qty = item.qty

        used_qty = min(available_qty, req_qty)

        difference = req_qty - used_qty

        # CHILD VALUES
        item.custom_requested_qty = req_qty


        item.custom_difference_qty = difference

        item.custom_fully_available = (
            req_qty if available_qty >= req_qty else 0
        )

        item.custom_partially_available = (
            used_qty if 0 < available_qty < req_qty else 0
        )

        # TOTALS
        requested_qty += req_qty

        available_qty_total += used_qty

        if available_qty >= req_qty:

            fully_available += req_qty

        elif available_qty > 0:

            partially_available += used_qty

        else:

            not_available += req_qty

    # PARENT TOTALS
    doc.custom_requested_sku = requested_sku

    doc.custom_requested_qty = requested_qty

    doc.custom_available_qty = available_qty_total

    doc.custom_difference_qty = (
        requested_qty - available_qty_total
    )

    doc.custom_fully_available = fully_available

    doc.custom_partially_available = partially_available

File: nafed_erp/nafed_erp/public/js/stock_entry.js
frappe.ui.form.on("Stock Entry", {
    refresh(frm) {
    console.log("🔄 Refresh triggered");
        

        frm.add_custom_button("Create PO From Indent", function() {

            frappe.call({
                method: "nafed_erp.procure_to_pay.api.stock_entry.create_po_from_indent",
                args: { stock_entry: frm.doc.name },
                callback: function(r) {
                    frappe.set_route("Form", "Purchase Order", r.message);
                }
            });

        });

    }
});

console.log("Stock Entry JS Loaded");

frappe.ui.form.on("Stock Entry", {

    onload(frm) {

        // Only for REPACK from GRN
        if (
            frm.doc.purchase_receipt_reference &&
            frm.doc.items.length === 0
        ) {

            frappe.db.get_doc(
                "Purchase Receipt",
                frm.doc.purchase_receipt_reference
            ).then(pr => {

                pr.items.forEach(row => {

                    let child = frm.add_child("items");

                    child.item_code = row.item_code;

                    child.qty = row.qty;

                    child.uom = row.uom;

                    child.s_warehouse = row.warehouse;

                    child.t_warehouse = row.warehouse;
                });

                frm.refresh_field("items");
            });
        }
    }
});

frappe.ui.form.on("Stock Entry", {

    refresh(frm) {

        calculate_indent(frm);

    }
});

frappe.ui.form.on("Stock Entry Detail", {

    item_code(frm, cdt, cdn) {

        calculate_indent(frm);

    },

    qty(frm, cdt, cdn) {

        calculate_indent(frm);

    },

    s_warehouse(frm, cdt, cdn) {

        calculate_indent(frm);

    },

    items_add(frm) {

        calculate_indent(frm);

    }
});

function calculate_indent(frm) {

    let requested_qty = 0;

    let available_qty_total = 0;

    let fully_available = 0;

    let partially_available = 0;

    let not_available = 0;

    frm.doc.items.forEach(function(row) {

        if (!row.item_code || !row.s_warehouse) return;

        frappe.call({

            method: "frappe.client.get_value",

            args: {

                doctype: "Bin",

                filters: {
                    item_code: row.item_code,
                    warehouse: row.s_warehouse
                },

                fieldname: ["actual_qty"]

            },

            callback: function(r) {

                let available_qty = r.message?.actual_qty || 0;

                let req_qty = row.qty || 0;

                let used_qty = Math.min(
                    available_qty,
                    req_qty
                );

                let difference = req_qty - used_qty;

                // CHILD VALUES
                frappe.model.set_value(
                    row.doctype,
                    row.name,
                    "custom_available_qty",
                    available_qty
                );

                frappe.model.set_value(
                    row.doctype,
                    row.name,
                    "custom_difference_qty",
                    difference
                );

                frappe.model.set_value(
                    row.doctype,
                    row.name,
                    "custom_fully_available",
                    available_qty >= req_qty ? req_qty : 0
                );

                frappe.model.set_value(
                    row.doctype,
                    row.name,
                    "custom_partially_available",
                    (available_qty > 0 && available_qty < req_qty)
                        ? used_qty : 0
                );

                // TOTALS
                requested_qty += req_qty;

                available_qty_total += used_qty;

                if (available_qty >= req_qty) {

                    fully_available += req_qty;

                } else if (available_qty > 0) {

                    partially_available += used_qty;

                } else {

                    not_available += req_qty;
                }

                // PARENT TOTALS
                frm.set_value(
                    "custom_requested_qty",
                    requested_qty
                );

                frm.set_value(
                    "custom_difference_qty",
                    requested_qty - available_qty_total
                );

                frm.set_value(
                    "custom_fully_available",
                    fully_available
                );

                frm.set_value(
                    "custom_partially_available",
                    partially_available
                );

            }

        });

    });

    frm.refresh_field("items");
}
